### PR TITLE
feat: [PL-32647]: Added support for pausing renewal of Vault connectors

### PR DESCRIPTION
**Pull request by tiwaritejas**
## Describe your changes
Renewal of Hashicorp vaults with token based auth will pause if the renewal fails for 2 times consecutively. Post this, renewal can resume in case the connector is updated or the manual test connection step succeeds by the user.

Testing:
1. Tested that the renewal gets paused in case the renewal fails for 2 times in a row.
2. Tested that the renewal is resumed when the connector is updated.
3. Tested that the renewal is resumed when the test connection is successful.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49671)

### CODE REVIEW

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 7 files at r1, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49671)* status: 7 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java` line 417 at r3](https://reviewable.io/reviews/harness/harness-core/49671#-NavLTcZ4fmmVR1qniJD:-NavLTcZ4fmmVR1qniJE:b1t02e) ([raw file](https://github.com/harness/harness-core/blob/9057fbca2ed99e12770c9fc5ad7293d31b7b91f9/120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java#L417)):*
> ```Java
>     Connector connector;
>     try {
>       connector = getConnectorWithIdentifier(accountIdentifier, orgIdentifier, projectIdentifier, identifier);
> ```

Can we instead use already existing get and update methods here?
___
*[`440-connector-nextgen/src/main/java/io/harness/connector/impl/NGVaultServiceImpl.java` line 1050 at r3](https://reviewable.io/reviews/harness/harness-core/49671#-NavNI181AnB2QNIV8jp:-NavNI181AnB2QNIV8jq:b-tu4dgi) ([raw file](https://github.com/harness/harness-core/blob/9057fbca2ed99e12770c9fc5ad7293d31b7b91f9/440-connector-nextgen/src/main/java/io/harness/connector/impl/NGVaultServiceImpl.java#L1050)):*
> ```Java
>     if (vaultConnector.getFailedRenewalAttempts() >= 10) {
>       log.warn(
>           "Stopping renewal iterator for vault- {} with id- {} as no successful renewal occurred in the last 3 hours",
> ```

Change message as attempts.

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49671)* status: 7 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java` line 417 at r3](https://reviewable.io/reviews/harness/harness-core/49671#-NavLTcZ4fmmVR1qniJD:-NbI_dq60h5VyIyGX06i:br9tgn0) ([raw file](https://github.com/harness/harness-core/blob/9057fbca2ed99e12770c9fc5ad7293d31b7b91f9/120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java#L417)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Can we instead use already existing get and update methods here?
</blockquote></details>

Existing update method involves many other operations like creation of heartbeat task, etc. which is not needed in this case. Hence, it is better to keep this flow separate.
___
*[`440-connector-nextgen/src/main/java/io/harness/connector/impl/NGVaultServiceImpl.java` line 1050 at r3](https://reviewable.io/reviews/harness/harness-core/49671#-NavNI181AnB2QNIV8jp:-NbI_sxWB0_7EVLsRzlk:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9057fbca2ed99e12770c9fc5ad7293d31b7b91f9/440-connector-nextgen/src/main/java/io/harness/connector/impl/NGVaultServiceImpl.java#L1050)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Change message as attempts.
</blockquote></details>

Done.

**Comment by tiwaritejas** [comment]
trigger messagecheck

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 3 of 4 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49671)* status: 7 file(s) to review. 1 of 2 approvals obtained.

**Comment by tiwaritejas** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 1]
```diff
@@ -409,6 +412,19 @@ public void updateConnectorEntityWithPerpetualtaskId(String accountIdentifier, S
...
       Long activityTime) {
+    Connector connector;
+    try {
+      connector = getConnectorWithIdentifier(accountIdentifier, orgIdentifier, projectIdentifier, identifier);
+    } catch (Exception ex) {
+      log.error(ex.getMessage());
+      return;
+    }
+    if (ConnectorType.VAULT.equals(connector.getType())) {
+      if (ConnectivityStatus.SUCCESS.equals(connectorValidationResult.getStatus())) {
```
Comment (gn-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:423]: nit: use &&

File: 440-connector-nextgen/src/main/java/io/harness/connector/impl/NGVaultServiceImpl.java [hunk 2]
```diff
@@ -172,10 +173,17 @@ public void renewToken(VaultConnector vaultConnector) {
...
         NGVaultRenewalTaskParameters.builder().encryptionConfig(baseVaultConfig).build();
 
-    DelegateResponseData delegateResponseData =
-        getDelegateResponseData(vaultConnector.getAccountIdentifier(), parameters, NG_VAULT_RENEW_TOKEN);
+    DelegateResponseData delegateResponseData;
+    try {
+      delegateResponseData =
+          getDelegateResponseData(vaultConnector.getAccountIdentifier(), parameters, NG_VAULT_RENEW_TOKEN);
+    } catch (Exception ex) {
+      pauseRenewalIfStale(vaultConnector);
```
Comment (gn-harness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/impl/NGVaultServiceImpl.java:181]: worthy of adding unit tests
 - when to pause
 - pausing logic, checking time buffers and all

Comment (tiwaritejas) [reply 440-connector-nextgen/src/main/java/io/harness/connector/impl/NGVaultServiceImpl.java:181]: Added unit tests

**Comment by tiwaritejas** [comment]
trigger uts

**Review by gn-harness — APPROVED** [review]

**Pull request merged**